### PR TITLE
Better HTTP/2 defaults.

### LINKED ISSUE #582: HTTP/2 should be off by default.
So...

I'm super excited that we've got an HTTP/2 implementation in here, but we also need to consider *what default behavior we should be aiming for*, and what we should prioritise.

Our HTTP/2 implementation is neccessarily going to be more complex, and less mature than our HTTP/1.1 implementation. Even if that *wasn't* the case, then HTTP/2 infrastructure that we're connecting too is *more* likely to have implementation issue and quirks than HTTP/1.1 versions.

I think we ought to clearly be *prioritizing for reliablity*, while still supporting users that want the more cutting-edge functionality.

I'd suggest that:

* We ought to have a documentation page on HTTP/2, with a brief overview of the primary differences, and motivation as to *why* user might or might not want to switch HTTP/2 support on.
* We ought to tidy up our interface for switching between HTTP versions.

We currently have this...

```python
client = Client(http_versions=["HTTP/1.1", "HTTP/2"])  # Really clunky
client = Client(http_versions=["HTTP/1.1"]) 
client = Client(http_versions=["HTTP/2"])  # Doesn't really make sense - you *might* still end up with an HTTP/1.1 connection.
```

I think we should just switch to...

```python
client = Client(http_2=True)
client = Client(http_2=False)
```

Furthermore, I think the default ought to be false. There's no point in exposing our users to this class of issues https://github.com/encode/httpx/issues?q=is%3Aissue+is%3Aopen+label%3Ahttp%2F2 *as our default behaviour*, but we can still make it easy for our users to upgrade to the HTTP/2 implementation.

### PR DESCRIPTION
**Pull request by tomchristie**
* HTTP/2 is off by default.
* `Client(http2=True)` to enable
* Added HTTP/2 docs.

Closes #582

Still TODO:

- [x] Interlink to the HTTP/2 section.
- [x] Link to https://http2-explained.haxx.se/content/en/ as a comprehensive HTTP/2 guide.

API changes for `http_version=...` to `http_2=[True|False]`, so we should leave this to a 0.9 release.

### CODE REVIEW

File: docs/http2.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+stream to handle multiple concurrent requests.
+
+HTTP/2 also provides support for functionality such as response prioritization,
+and server push.
+
+## Enabling HTTP/2
+
+The HTTPX client provides provisional HTTP/2 support.
+
+The HTTP/2 support is not enabled by default, because HTTP/1.1 a mature,
```
Comment (florimondmanca) [inline docs/http2.md]: Typo: *is* a mature

File: docs/http2.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+battle-hardened transport layer. With HTTP/2 being newer and significantly more
+complex, it should be considered a less robust option at this point in time.
+
+However, if you're issuing highly concurrent requests you might want to consider
+trying out our HTTP/2 support. You can do so by instantiating a client with
+HTTP/2 support enabled.
+
+Simply instantiating a client...
```
Comment (florimondmanca) [inline docs/http2.md]: This line is probably superfluous — how about adding a colon at the end of the sentence above?

Also let’s not assume things are « simple »? 😄 (I remember someone did tremendous work to remove these expressions from the Django docs last month to make them more inclusive, but I can’t find the PR/tweet anymore).

File: docs/http2.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+```python
+client = httpx.Client(http_2=True)
+...
+```
+
+Instantiating a client as a context manager, to ensure that all connections
+are nicely closed at the end of it's usage...
```
Comment (florimondmanca) [inline docs/http2.md]: Typo: its?

File: docs/http2.md (hunk 1 above)
Comment (tomchristie) [reply docs/http2.md]: Thanks!

File: docs/http2.md (hunk 2 above)
Comment (tomchristie) [reply docs/http2.md]: Yes, and yes!

File: docs/http2.md (hunk 3 above)
Comment (tomchristie) [reply docs/http2.md]: Going to change that around in any case, yeah. Thanks again!

**Comment by tomchristie** [comment]
> I suppose this moves us forward relative to minimal dependencies, eg we’ll now be able to not require h2 unless users want to use HTTP/2?

Ah yeah I'd forgotten about that. We *can also* make that switch here, yup.
I guess we would drop `h2` out of our default install, and into `extra_requires`, so...

`pip install httpx[h2]`

And a check on the client for "if http_2 is True and h2 is not installed then error".

**Comment by florimondmanca** [comment]
> We can also make that switch here, yup.

I’d rather we deal with it later, this PR is quite a big chunk already. :-)

**Comment by tomchristie** [comment]
Good call.

**Comment by tomchristie** [comment]
Should we use `httpx.Client(http_2=True)` or `httpx.Client(http2=True)`?

I think the later is probably neater, right?

**Comment by florimondmanca** [comment]
Just remembered we had a "Minimal installs" issue here — https://github.com/encode/httpx/issues/496.

**Pull request merged**